api/routes/application.py

```diff
@@ -77,37 +85,94 @@ def create_application(
     return application
 
 
-@application_router.put("/{application_id}", status_code=status.HTTP_200_OK, response_model=s.ApplicationOut)
+@application_router.put(
+    "/{application_id}",
+    status_code=status.HTTP_200_OK,
+    response_model=s.ApplicationPutOut,
+    responses={
+        status.HTTP_404_NOT_FOUND: {"description": "Application not found"},
+    },
+)
 def update_application(
     application_id: int,
-    data: s.ApplicationPut,
+    data: s.ApplicationPutIn,
     db: Session = Depends(get_db),
     current_user: m.User = Depends(get_current_user),
 ):
-    application: m.Application | None = db.scalar(sa.select(m.Application).where(m.Application.id == application_id))
+    if not data.status:
+        log(log.ERROR, "Status is required")
+        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Status is required")
+
+    application: m.Application | None = db.scalar(
+        sa.select(m.Application).where(m.Application.is_deleted.is_(False), m.Application.id == application_id)
+    )
 
     if not application:
         log(log.ERROR, "Application [%s] not found", application_id)
         raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Application not found")
-    job_owner_id: int | None = db.scalar(sa.select(m.Job.owner_id).where(m.Job.id == application.job_id))
-    if not job_owner_id:
+
+    if application.status in [m.ApplicationStatus.ACCEPTED, m.ApplicationStatus.REJECTED]:
+        log(log.ERROR, "Application [%s] is already completed", application_id)
+        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Application is already completed")
+
+    job = db.scalar(sa.select(m.Job).where(m.Job.id == application.job_id))
```

Comment:
application should have relationship on job. No need to do extra DB query here, just do `application.job...`